docs/tests.md

```diff
@@ -0,0 +1,454 @@
+  ser -- expected output --> oc;
+```
+
+1. **KAT Generator Functions:** We utilize KAT generator functions located in
+   the `tests/external_estimators` directory. These functions have hardcoded
+   inputs as they are not intended for regular modification. Both input
+   parameters and their corresponding outputs (referred to as `expected_outputs`
+   as they act as reference values) are serialized into the `tests/kat.yaml`
+   file for later use during estimator testing.
+
+2. **Internal Estimation Functions:** With a collection of inputs and their
+   expected outputs, we define how these inputs should be processed within our
+   library. This is achieved through internal estimation functions found in the
+   `tests/internal_estimators` directory. Each function corresponds to a KAT
+   generator in `tests/external_estimators`.
+
+3. **Test Execution and Comparison:** In the final step, we execute all our
+   internal estimation functions using the serialized inputs from
+   `test/kat.yaml`. The calculated outputs are then compared against the
+   expected outputs from the KAT generators to verify the correctness and
+   accuracy of our library's implementations.
+
+This separation of concerns makes it easier to maintain and extend the test
+suite as we introduce new estimators or modify existing ones.
+
+### Writing Tests: A Step-by-Step Guide
+
+This guide walks you through incorporating a new KAT test for a library
+estimator, offering insights relevant for both understanding and updating
+existing tests.
+
+#### 1. Define Your KAT Generator Function
+
+This step involves defining or declaring the KAT generator functions which
+produces the expected outputs from hardcoded inputs. These functions require
+careful crafting and may originate from external sources or researchers.
+
+- Begin by creating a new file named `ext_<your_new_estimator>` within the
+  `tests/external_estimators/` directory. This file can be written in either
+  Sage or Python, as our framework supports both formats.
+
+- Inside this file, define your KAT generator function using the naming
+  convention `ext_<kat_gen_function_name>`. These functions contain the logic to
+  calculate expected outputs for their predefined inputs.
+
+For example, consider the `ext_lee_brickell` function:
+
+```python
+#Defined at ext_sdfq.py
+def ext_lee_brickell():
+    """Generate expected complexities for Lee-Brickell SDFq problems.
+
+    This function calculates the expected complexities for a predefined set of
+    Lee-Brickell SDFq problem parameters.
+
+    Returns:
+        list of tuple: Each tuple contains:
+            - tuple: Input parameters (n, k, w, q)
+            - float: Corresponding expected complexity
+    """
+
+    inputs = [(256r, 128r, 64r, 251r), (961r, 771r, 48r, 31r)]
+
+    def gen_single_kat(input: tuple):
+        n, k, w, q = input
+        expected_complexity = #necessary calculations to produce the expected complexity
```

Comment:
expected_complexity = #necessary calculations to produce the expected complexity ---> expected_complexity = output of the external estimating the complexity of the lee-brickell algorithm on inputs `(n, k, w, q)`

Reply:
Applied.